src/main/java/com/sonyericsson/hudson/plugins/gerrit/trigger/workflow/GerritReviewStep.java

```diff
@@ -0,0 +1,144 @@
+import org.jenkinsci.plugins.workflow.steps.StepContextParameter;
+import org.kohsuke.stapler.DataBoundConstructor;
+import org.kohsuke.stapler.DataBoundSetter;
+
+/**
+ * Allows altering the Gerrit review posted at the end of build during the build.
+ * @author Teemu Murtola &lt;teemu.murtola@gmail.com&gt
+ */
+public class GerritReviewStep extends AbstractStepImpl {
+
+    private String customUrl;
+    private String unsuccessfulMessage;
+
+    /**
+     * Constructor.
+     *
+     * There are no mandatory parameters to the step.
+     */
+    @DataBoundConstructor
+    public GerritReviewStep() {
+    }
+
+    /**
+     * Gets the custom URL for a step.
+     * @return the URL.
+     */
+    @CheckForNull
+    public String getCustomUrl() {
+        return customUrl;
+    }
+
+    /**
+     * Sets a custom URL to post for a build.
+     * @param customUrl the URL to post.
+     */
+    @DataBoundSetter
+    public void setCustomUrl(String customUrl) {
+        this.customUrl = Util.fixEmptyAndTrim(customUrl);
+    }
+
+    /**
+     * Gets the unsuccessful message for a step.
+     * @return the message.
+     */
+    @CheckForNull
+    public String getUnsuccessfulMessage() {
+        return unsuccessfulMessage;
+    }
+
+    /**
+     * Sets additional information to post for a failed build.
+     * @param unsuccessfulMessage Additional information to post for a failed build.
+     */
+    @DataBoundSetter
+    public void setUnsuccessfulMessage(String unsuccessfulMessage) {
+        this.unsuccessfulMessage = Util.fixEmptyAndTrim(unsuccessfulMessage);
+    }
+
+    /**
+     * Executes the GerritReviewStep.
+     */
+    public static class Execution extends AbstractSynchronousStepExecution<Void> {
+
+        @StepContextParameter
+        private transient Run build;
+
+        @Inject
+        private transient GerritReviewStep step;
+
+        @Override
+        protected Void run() throws Exception {
+            ToGerritRunListener listener = ToGerritRunListener.getInstance();
+            String customUrl = step.getCustomUrl();
+            if (customUrl != null) {
+                listener.setBuildCustomUrl(build, customUrl);
+            }
+            String unsuccessfulMessage = step.getUnsuccessfulMessage();
+            if (unsuccessfulMessage != null) {
+                listener.setBuildFailureMessage(build, unsuccessfulMessage);
+            }
+            return null;
+        }
+
+        private static final long serialVersionUID = 1L;
+    }
+
+    /**
+     * Adds the step as a workflow extension.
+     */
+    @Extension(optional = true)
+    public static class DescriptorImpl extends AbstractStepDescriptorImpl {
+
+        /**
+         * Constructor.
+         */
+        public DescriptorImpl() {
+            super(Execution.class);
+        }
+
+        @Override
+        public String getFunctionName() {
+            return "gerritReview";
```

Comment:
I agree that the run() implementation would be very different, probably for all four cases. But that is relatively easy to dispatch to different functions with just a few if's. The other side is that if it is in different steps, it may end up duplicating a lot of the setters/getters for the step parameters (at least most of which would/should be identical for all the steps to keep things consistent), and the jelly/html files for them. I don't know enough of how the jelly side works to know whether there is some way to avoid that. But that said, I'm fine with splitting this, and I agree that if we split it, we should rename the current implementation.

Another, minor concern that I have in the back of my head is that if there are a lot of plugins each contributing a lot of different workflow steps, at least the current workflow snippet generator might not be very good for finding them. But that might not be a concern for this change...